### PR TITLE
[CTF Rebuild] Siege rework. Catapult launch force changes. Added outposts

### PR DESCRIPTION
**Pull request by bunniewormy**
## Status

(pick one)

- **READY**: this PR is (to the best of your knowledge) ready to be incorporated into the game.

## Description

[change] catapult changes:
		players are launched with 75% original force
		other items are launched with 110% original force
[change] cata rocks do more damage to blocks 
[added] cata rocks are shieldable
[removed] spawning removed from ballista 
[change] balllista no longer costs gold
[change] ballista price reduced to 150
[change] improved cata unpacking code 
[added] added ballista bomb bolts to tth
[added, big addition] add an outpost to the siege shop. Outpost mechanics:
		5x5 outpost which is essentially a workshop, but is unpacked from a crate bought in the siege shop for 150 coins and 50 gold
        respawning with shorter spawn immunity
        storage-size storage
        can change class on it (both teams)
        also works as a tunnel (only owner team can use the tunnel; other teams can however use tunnel to travel to it)
        capturable same as ballistas

### CODE REVIEW

**Comment by MonkeyFeats** [comment]
The outpost seems like a big change, it would require extensive testing to evaluate whether it is a good change or just pointless.

**Comment by bunniewormy** [comment]
> 
> 
> The outpost seems like a big change, it would require extensive testing to evaluate whether it is a good change or just pointless.

The point of the change is for ballistas to actually be a siege machine rather than a spawn point, and for outposts to be spawn points instead of them. It doesn't make much sense for something that is supposed to shoot bolts to be a respawn point. Moreover adding an improved respawn point with a tunnel improves builders' ability to control terrain + the storage is pretty useful for this as well.

It's not that big of a change - the outposts don't really do anything that other structures in the game don't, it's just ballistas with tunnels instead of ability to shoot.

File: Entities/Industry/CTFShops/Outpost/Outpost.as [hunk 1]
```diff
@@ -0,0 +1,83 @@
+﻿// Vehicle Workshop
+
+#include "GenericButtonCommon.as"
+#include "StandardRespawnCommand.as"
+#include "StandardControlsCommon.as"
+#include "TunnelCommon.as"
+
+u8 respawn_immunity_time = 30 * 1; // 1 second
```
Comment (Vam-Jam) [inline Entities/Industry/CTFShops/Outpost/Outpost.as:8]: const

File: Entities/Items/Crate/Crate.as [hunk 2]
```diff
@@ -441,7 +423,30 @@ void Unpack(CBlob@ this)
...
+
+		if (map.rayCastSolid(this.getPosition(), this.getPosition() + Vec2f(0, 14.0f), adjusted_pos))
+		{
+			adjusted_pos -= Vec2f(0, blob.getHeight() / 2 + 4.0f);
+		}
+		else
+		{
+			if (blob.getName() == "outpost")
+			{
+				//pos -= Vec2f(0, map.tilesize);
```
Comment (Vam-Jam) [inline Entities/Items/Crate/Crate.as:444]: is this suppose to be commented out, if so remove the if empty if statement

File: Entities/Vehicles/Catapult/Catapult.as [hunk 3]
```diff
@@ -233,10 +236,18 @@ void Vehicle_onFire(CBlob@ this, VehicleInfo@ v, CBlob@ bullet, const u8 _charge
...
+
 		vel.RotateBy(angle);
 
-		bullet.setVelocity(vel);
+
+		if(bullet.hasTag("player"))
+		bullet.setVelocity(vel * player_launch_modifier);
+
+		else
+		bullet.setVelocity(vel * other_launch_modifier);
```
Comment (Vam-Jam) [inline Entities/Vehicles/Catapult/Catapult.as:250]: fix formatting

File: Rules/CommonScripts/RespawnSystem.as [hunk 4]
```diff
@@ -33,6 +33,11 @@ shared class RespawnSystem
 			CBlob @newBlob = server_CreateBlob(p_info.blob_name, p_info.team, at);
 			newBlob.server_SetPlayer(player);
 			player.server_setTeamNum(p_info.team);
+
+			if(p_info.customImmunityTime > 0)
```
Comment (Vam-Jam) [inline Rules/CommonScripts/RespawnSystem.as:37]: if space (

File: Entities/Items/Crate/Crate.as [hunk 5]
```diff
@@ -184,8 +184,7 @@ void Land(CBlob@ this)
 
 bool doesCollideWithBlob(CBlob@ this, CBlob@ blob)
 {
-	return (this.getName() == blob.getName())
-	    || ((blob.getShape().isStatic() || blob.hasTag("player") || blob.hasTag("projectile")) && !blob.hasTag("parachute"));
+	return (blob.getShape().isStatic() || blob.getPlayer() !is null || blob.hasTag("projectile")) && !blob.hasTag("parachute");
```
Comment (CatWiz) [inline Entities/Items/Crate/Crate.as:187]: this prevents crates from colliding with each other again

File: Entities/Items/Crate/Crate.as [hunk 6]
```diff
@@ -739,7 +758,7 @@ CBlob@ getPlayerInside(CBlob@ this)
 	return null;
 }
 
-bool DumpOutItems(CBlob@ this, float pop_out_speed = 5.0f, Vec2f init_velocity = Vec2f_zero, bool dump_special = true)
+bool DumpOutItems(CBlob@ this, float pop_out_speed = 5.0f, Vec2f init_velocity = Vec2f_zero, bool dump_player = true)
```
Comment (CatWiz) [inline Entities/Items/Crate/Crate.as:761]: looks like there's a bunch of unintentional changes to mine trap code

File: Entities/Industry/CTFShops/Outpost/Outpost.as (hunk 1 above)
Comment (bunniewormy) [reply Entities/Industry/CTFShops/Outpost/Outpost.as]: fixed

File: Entities/Vehicles/Catapult/Catapult.as (hunk 3 above)
Comment (bunniewormy) [reply Entities/Vehicles/Catapult/Catapult.as]: fixed

File: Rules/CommonScripts/RespawnSystem.as (hunk 4 above)
Comment (bunniewormy) [reply Rules/CommonScripts/RespawnSystem.as]: fixed

File: Entities/Items/Crate/Crate.as (hunk 2 above)
Comment (bunniewormy) [reply Entities/Items/Crate/Crate.as]: fixed

File: Entities/Items/Crate/Crate.as (hunk 5 above)
Comment (bunniewormy) [reply Entities/Items/Crate/Crate.as]: fixed

File: Entities/Items/Crate/Crate.as (hunk 6 above)
Comment (bunniewormy) [reply Entities/Items/Crate/Crate.as]: fixed

File: Entities/Items/Crate/Crate.as [hunk 7]
```diff
@@ -348,27 +348,10 @@ void onCommand(CBlob@ this, u8 cmd, CBitStream @params)
...
-				CBlob@ item = inv.getItem(i);
-				if (item.getName() == "mine" && item.getTeamNum() != caller.getTeamNum())
-				{
-					CBitStream params;
-					params.write_u16(caller.getNetworkID());
-					params.write_u16(item.getNetworkID());
-					this.SendCommand(this.getCommandID("boobytrap"), params);
-					return;
-				}
-			}
```
Comment (CatWiz) [inline Entities/Items/Crate/Crate.as:369]: why did you just remove this?

Comment (bunniewormy) [reply Entities/Items/Crate/Crate.as]: Fixed, accident

**Comment by mugg91** [comment]
If you remove spawning from Ballista, might as well remove it from Warboat, too.

> also works as a tunnel (only owner team can use the tunnel; other teams can however use tunnel to travel to it)

If the other team can travel there, then that means the other team can just travel in with a lit keg to destroy it.

**Comment by AsuMagic** [comment]
> If you remove spawning from Ballista, might as well remove it from Warboat, too.

Maybe, but this would make warboats pretty useless compared to longboats, as warboats suck for ramming. Either way, they are entirely unused in CTF and TTH is a bit of a second class citizen, so I don't think it is very important to change this.

**Comment by AsuMagic** [comment]
Note that I have reworked catapult rocks a bit, which was merged into this PR. See: https://github.com/bunniewormy/Base/pull/47

> - Fine-tuned tile and blob damage a fair bit (stone damage per rock significantly increased, wood damage per rock increased even further)
> - Hopefully made rock damage more consistent than before regardless of shooting distance/force/luck
> - Visual particle overhaul (wooden particles on hitting wood, dust particles on hitting ground, stone dust particles in flight, smaller gib particle when the rock cata is destroyed) and minor audio tweaks
> - Ricochet happens quite a lot more and deals more damage due to some physics tweaks, which imo feels good, and makes cata a bit deadlier in some circumstances: e.g. behind just in front of a tower under cata fire, or being inside of a tower as rocks fall into it
> - Catapults fire 7 rocks instead of 5 (40% increase in rocks per catapult shot)
> - Rock cost is increased from 2 to 3 stone (50% more expensive)

It was tested both offline and online with no major observed issues, including with lag emulation.

**Review by AsuMagic — CHANGES_REQUESTED** [review]
General observations or issues I've observed aside of the other comments:
- The ballista icon still shows up on the minimap, even though they aren't as critical as before (catapults don't show). Outposts don't. This may be caused by `Spawner.as` which `Ballista.cfg` uses but not `Outpost.cfg`.
- We should test other gamemodes to see if any of the respawn/ballista changes negatively affects them. TTH, for one, probably is...

File: Entities/Items/Crate/Crate.as [hunk 8]
```diff
@@ -185,7 +185,7 @@ void Land(CBlob@ this)
 bool doesCollideWithBlob(CBlob@ this, CBlob@ blob)
 {
 	return (this.getName() == blob.getName())
-	    || ((blob.getShape().isStatic() || blob.hasTag("player") || blob.hasTag("projectile")) && !blob.hasTag("parachute"));
+		|| ((blob.getShape().isStatic() || blob.getPlayer() !is null || blob.hasTag("projectile")) && !blob.hasTag("parachute"));
```
[the remark this answers is not in the document]
Comment (bunniewormy) [reply Entities/Items/Crate/Crate.as]: i dont know
reverted

File: Entities/Common/Includes/Costs.as [hunk 9]
```diff
@@ -139,6 +139,8 @@ void InitCosts()
 	CTFCosts::ballista_gold =               ReadCost(cfg, "cost_ballista_gold"              , 50);
```
[the remark this answers is not in the document]
Comment (bunniewormy) [reply Entities/Common/Includes/Costs.as:139]: ok

File: Rules/CommonScripts/RespawnSystem.as [hunk 10]
```diff
@@ -33,6 +33,11 @@ shared class RespawnSystem
 			CBlob @newBlob = server_CreateBlob(p_info.blob_name, p_info.team, at);
 			newBlob.server_SetPlayer(player);
 			player.server_setTeamNum(p_info.team);
+
+			if (p_info.customImmunityTime > 0)
```
[the remark this answers is not in the document]
Comment (bunniewormy) [reply Rules/CommonScripts/RespawnSystem.as]: done

**Comment by bunniewormy** [comment]
removed ballista's minimap icon
replaced ballista minimap icon with outpost icon in the spritesheet 
removed Spawner.as from Ballista.cfg script list 
added Spawner.as to Outpost.cfg script list; removed createRespawnPoint() from Outpost.as file

**Review by AsuMagic — APPROVED** [review]

**Pull request merged**